### PR TITLE
In-app code review

### PR DESCRIPTION
**Pull request by Safi-Ullah**
## Issue:
https://github.com/fenske/devskills-app/issues/239

## What New: 
- Separate panel for `Code Review` and `Hiring decision`.
- Expandable sections.
- Define the scoring table tooltip in the app
- HTML-rendered tooltips
- Overall avg of code review

### CODE REVIEW

**Comment by Safi-Ullah** [comment]
Created the PR just to test a few things on the staging server. Tooltips are still pending, hence the `WIP` tag.

**Comment by Safi-Ullah** [comment]
Since, a new `Code review` panel is added, should I remove this one and add the link to the new one.
Screenshot 2021-07-14 at 12 55 04 AM

**Comment by fenske** [comment]
> Since a new `Code review` panel is added, should I remove this one and add the link to the end of new one.
> Screenshot 2021-07-14 at 12 55 04 AM

Good that you asked. 🙂 I didn't specify it in the original AC, but this panel should render for all candidates invited before the new in-app code review panel has been introduced. Otherwise, users won't have a chance to locate the corresponding scorecards.

**Comment by Safi-Ullah** [comment]
> this panel should render for all candidates invited before the new in-app code review panel has been introduced. Otherwise, users won't have a chance to locate the corresponding scorecards.

Oh, oka...so this shouldn't show for those who're invited afterward(have a scorecard in their doc).

**Comment by fenske** [comment]
> > this panel should render for all candidates invited before the new in-app code review panel has been introduced. Otherwise, users won't have a chance to locate the corresponding scorecards.
> 
> Oh, oka...so this shouldn't show for those who're invited afterward(have a scorecard in their doc).

Exactly. 💯 

Btw, I'll start a review in the meantime. Likely will take a bit, given the size of the change. Will share my findings once I'm done.

**Comment by Safi-Ullah** [comment]
> Btw, I'll start a review in the meantime. Likely will take a bit, given the size of the change. Will share my findings once I'm done.

Yeah, sure.

**Comment by fenske** [comment]
@Safi-Ullah Let's separate the review into two parts:

1. UI/UX only
2. Code review

Let's start with UI/UX. Once we are clear on that, I'll do a separate code review.

I'll list all points I thought of, please check them off as you fix them. ✔️

Also, if you disagree with anything or need clarification, as usual, just let me know.

So here we go.

### UI/UX comments

#### Code review panel

- [ ] The total review score badge seems missing (see the [mockup](https://user-images.githubusercontent.com/1162212/124760555-2578d700-df31-11eb-94e8-c649633a3ad4.png)).
- [ ] When deciding on selecting the rating for a particular competence, the stars should fill with the indigo color with a reduced opacity on hover to give a sense that they're editable.
- [ ] The unselected stars should have the border of a shadow of grey. Too much contrast with black.
- [ ] The HTML tooltips don't seem to render HTML properly (e.g. check how the SQLite requirement is stored in Firestore and how it's rendered)
- [ ] The scoring criteria tooltip should be 1.5x wider than the current width. Otherwise, it's hard to see on smaller screens.
- [ ] It's hard to figure that code review sections are expandable. Add the chevron-up icon on the left (just before each requirement ) that would switch to chevron-down when the corresponding section is expanded. See [example](https://headlessui.dev/react/disclosure).
- [ ] Requirement comment text box: the font color should turn black when editing.

### Hiring decision panel
- [ ] For the "Advance the candidate" button group, the font of the selected button should turn indigo as the border.
- [ ] Final comments text box: the font color should turn grey when not editing.
- [ ] Let's rename "Final Comments" to "Final comments" for casing consistency.
- [ ] "Advance the candidate to the next round" → "Advance the candidate to the next round?"

#### Misc
- [ ] All Edit buttons should look as secondary (similar to Cancel) as they are secondary actions.
- [ ] Reduce the font size a bit in all text boxes.


(Really looking forward to releasing this! It's going to make a huge impact on saving hiring teams' time! 🤩 )

File: components/CriteriaLineItem.tsx [hunk 1]
```diff
@@ -35,9 +35,9 @@ const CriteriaLineItem = ({ requirement, criteria, score, comments, handleScorec
                 </span>
                 <span className="px-1" onClick={e => e.stopPropagation()}>
                   <CustomPopover icon={<InfoIcon />}>
-                    <div className="overflow-hidden rounded-lg bg-white w-4/6 mx-40 p-4 shadow-lg ring-1 ring-black ring-opacity-5">
+                    <div className="overflow-hidden rounded-lg bg-white p-4 shadow-lg ring-1 ring-black ring-opacity-5">
                       <h3 className="text-lg leading-6 font-medium text-gray-900 pb-4">Scoring criteria</h3>
-                      <span className="leading-7 whitespace-pre-wrap" dangerouslySetInnerHTML={{ __html: criteria }} />
+                      <span className="prose sm:prose-sm prose-indigo" dangerouslySetInnerHTML={{ __html: criteria }} />
```
Comment (fenske) [inline components/CriteriaLineItem.tsx:40]: The prose classes give HTML tooltips a nicer look.

CleanShot 2021-07-14 at 22 20 46

File: components/CandidateHiringDecisionPanel.tsx [hunk 2]
```diff
@@ -60,7 +60,10 @@ const CandidateHiringDecisionPanel = ({ hiringDecision={ comments: '', decision:
                 isEditing
                   ? <div className="flex-shrink-0 py-4 space-x-4 flex justify-end">
                     <span className="inline-flex rounded-md shadow-sm">
-                      <button type="button" onClick={() => setIsEditing(false)}
+                      <button type="button" onClick={() => {
+                        reset([comments]);
```
Comment (fenske) [inline components/CandidateHiringDecisionPanel.tsx:64]: Resetting fixes the following problem with text that wasn't saved but still kept in the text area.

CleanShot 2021-07-14 at 21 45 31

File: components/CriteriaLineItem.tsx (hunk 1 above)
Comment (Safi-Ullah) [reply components/CriteriaLineItem.tsx:40]: Oh, oka 😮....I thought the issue was with `spacing` in html rendering.

Comment (fenske) [reply components/CriteriaLineItem.tsx:40]: That's what I meant with prose classes 😃

CleanShot 2021-07-15 at 11 09 49

Comment (Safi-Ullah) [reply components/CriteriaLineItem.tsx:40]: yes, right...I missed this.

**Comment by fenske** [comment]
@Safi-Ullah I've completed my code review. 

I also pushed some changes which pretty much represent my feedback (I had a bit of spare time so wanted to avoid back-and-forth, hence pushed directly).

Please check it out and let me know if those changes make sense to you.

**Comment by Safi-Ullah** [comment]
Yeah, looks good.

I see that you've moved a couple of components outside the library...Is this solely decided based on where it's being used? Because I thought that refactored/util components are included here.

**Comment by fenske** [comment]
> Yeah, looks good.
> 
> I see that you've moved a couple of components outside the library...Is this solely decided based on where it's being used? Because I thought that refactored/util components are included here.

Btw, nothing is set in stone. 😃 Anyway, my thinking was to have a logical separation between the different app parts namely:

1. library - anything only used for the challenge library
2. candidateList - anything only used for the candidates list

etc.

So since both Panel and CustomPopover were used by the components that belonged to the candidateList (and it looked like they could be re-used in other parts of the app later), I moved them out of library.

**Comment by Safi-Ullah** [comment]
> 1. library - anything only used for the challenge library

Oka, the term confused me. I had this impression in my mind that this was a library of reusable(utility) components.

**Comment by fenske** [comment]
> > 1. library - anything only used for the challenge library
> 
> 
> 
> Oka, the term confused me. I had this impression in my mind that this was a library of reusable(utility) components.

Agree its confusing. Might need a better name.

**Pull request merged**